Preserve category in merged manifest

epananth opened the pull request:
Issue -> https://github.com/dotnet/arcade/issues/6517

This is part -1 pr for preserving the category of blobs in merged manifest in arcade-services 

So if no category is present then I added empty string. I think this is not a good idea. Its better to add a new category like 'NONE' or something here. 

So when we infer category here in arcade, I can check if its 'NONE' then go determine the category like we used to before https://github.com/dotnet/arcade/blob/6818f9bff65fb7277174876bdf207f5ac8933286/src/Microsoft.DotNet.Build.Tasks.Feed/src/PublishArtifactsInManifestBase.cs#L1106

else if its other then do something else. 

This is how the merged manifest looks like now
![Capture](https://user-images.githubusercontent.com/47157394/141875423-647853d4-77a8-4d9c-b689-3bc7269266a8.PNG)

Arcade test -> https://dev.azure.com/dnceng/internal/_build/results?buildId=1470781&view=results

riarenas commented:
> So if no category is present then I added empty string. I think this is not a good idea. Its better to add a new category like 'NONE' or something here.

I think I'm missing something, I don't see this NONE category added anywhere in this PR?

mmitche commented on src/Maestro/Client/src/Generated/Models/AssetData.cs:
```diff
@@ -22,5 +22,8 @@ public AssetData(bool nonShipping)
         public IImmutableList<Models.AssetLocationData> Locations { get; set; }
+
+        [JsonProperty("category")]
+        public string Category { get; set; }
```
We can't modify the AssetData here. When we regenerate the model, it will remove this property

mmitche replied:
So basically I think we need to simply not use AssetData as the intermediate type that is used to construct the merged manifest.

epananth replied:
I hate me :( I will fix this..

epananth replied:
I fixed this

epananth commented:
So I updated this code NOT to use generated code, instead I stored that info ConditionalWeakTable.

Here is the arcade test -> https://dev.azure.com/dnceng/internal/_build/results?buildId=1473440&view=results

![Capture](https://user-images.githubusercontent.com/47157394/142273535-42cf4b93-0edc-406d-a3bd-584480d080c3.PNG)

epananth commented:
> > So if no category is present then I added empty string. I think this is not a good idea. Its better to add a new category like 'NONE' or something here.
> 
> I think I'm missing something, I don't see this NONE category added anywhere in this PR?

I did not add the category to NONE, I thought it will be a good idea.. I added that in this run also included the test and screenshot of the merged manifest

mmitche commented on src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs:
```diff
@@ -42,6 +43,10 @@ public class PushMetadataToBuildAssetRegistry : MSBuild.Task, ICancelableTask
 
+        public string AssetCategory { get; set; }
+
+        public ConditionalWeakTable<AssetData, string> assetWithCategory = new ConditionalWeakTable<AssetData, string>();
```
Should this be public?

mmitche replied:
Also, what's the reasoning for using a `ConditionalWeakTable` vs. a normal dictionary?

mmitche replied:
Also, can you add some comments here as to why we're using a side-dictionary to maintain the category? I think this solution is okay, though I think it would be 'cleaner' to have the flow be the following:

1. The assets, signing info are merged, etc. are merged using the **original** types (PackageArtifactModel/BlobArtifactModel).
2. The merged xml manifest is created and pushed to artifacts
3. The Maestro build model is created using this merged data.

That would eliminate the need for side-table information and make the implementation cleaner, though it would be a decent amount of additional code change. Right now the use of AssetData and BuildData is a bit odd. We're basically starting as a model based on the xml manifest, constructing the merged manifest using the Maestro model of build data, then attempting to go back to the model based on the xml manifest. which is a little awkward. It worked fine because the models were close enough to each other to work in most scenarios, but it wasn't ideal.

epananth replied:
I was slightly hesitant to change the current functionality, so I added a the fix with the table. But I don't want to add more technical debt, so I will try an clean this up. Thanks for clarifying this!

mmitche commented on src/Maestro/Microsoft.DotNet.Maestro.Tasks/src/PushMetadataToBuildAssetRegistry.cs:
```diff
@@ -42,6 +43,10 @@ public class PushMetadataToBuildAssetRegistry : MSBuild.Task, ICancelableTask
 
         public string AssetVersion { get; set; }
 
+        public string AssetCategory { get; set; }
```
What's the reason for adding this as a task parameter?

epananth commented:
Closing this PR, going to open another PR for this issue
